### ISSUE TITLE
Split TelemetryInfo fields to separate diagnostics

### ISSUE THREAD

**Issue by eakuefner** [catapult-project/catapult#3507]
We're going to have benchmarks that are not Telemetry-based, and in these cases, we'll still need `benchmarkName` and `benchmarkStart`, and maybe also `label`. We should probably split those into a new `BenchmarkInfo`. We might also want to do something with the rest of the fields (StoryInfo?) and figure out to what extent we actually need tir_label.

We can deal with this for development purposes, but I think this blocks productionizing the histogram pipeline because we can't start storing TelemetryInfos in the datastore if TelemetryInfo won't continue to exist as-is.

@benshayden @simonhatch

**Comment by benshayden** [comment]
Updating this bug based on "Rationalizing RevisionInfo and TelemetryInfo" with steps for the TBMv2 side. (Should this bug also track the dashboard side?)

- Merge TelemetryInfo with MergedTelemetryInfo: #3607
- Split benchmarkStart to tr.v.d.DateRange named "testSuiteStarts" #3608
- Split canonicalUrls to tr.v.d.UrlSet named "traceUrls"
- Split benchmarkNames to tr.v.d.Generic named "testSuiteNames"
- Split labels to tr.v.d.Generic named "resultsLabels"
- Split story names, tags, grouping keys, TIRs, repetition counts to tr.v.d.StoryInfo named "stories" with plural fields for mergeability

**Comment by simonhatch** [comment]
Dashboard side seems minimal (think it's just a few references in /add_histogram* , @eakuefner correct me if I'm wrong), would just tack a step on to update dashboard side.

**Comment by eakuefner** [comment]
@simonhatch You're correct. In fact, the only thing that needs to be updated as a result of this is the list of sparse diagnostic types in add_histogram.py. I spoke offline to @benshayden, and to unblock deduplicating sparse diagnostics without requiring this refactor to be complete, I'll just add `__eq__` methods to the existing sparse diagnostic types. The code that actually uses this equality checking should not require any modification in the course of doing this refactor, because of duck typing.

**Comment by benshayden** [comment]
I started doing this and I noticed that Generic (actually, its merged variant, CollectedGeneric) collects duplicate entries in an array. We actually want a Set that deduplicates. Most of these fields are strings (storyset repetition counter is a number, but is never used for arithmetic). The URLs (canonical trace url and log url) are strings; string-set-span can display strings as links if they match URL_REGEX, which is what generic-object-view already does.

Splitting TelemetryInfo will unblock deduplicating SparseDiagnostic and fix #3582 and #3293 .
BuildbotInfo and DeviceInfo also need to be split for similar reasons #3294 
Ownership can be split into 2 StringSets in order to reduce code duplication.

I considered allowing users to cap the number of strings merged into a StringSet, but I'm not sure that we need that complexity yet.

@simonhatch Any questions before I send you a CL to implement a StringSet Diagnostic?

**Comment by simonhatch** [comment]
I'll grab you on vc tomorrow to get some more context here.

**Comment by benshayden** [comment]
VC SGTM.

Note to self: temporary code for back-compat?

**Comment by benshayden** [comment]
Ownership is not currently mergeable. Merging Ownerships produces UnmergeableDiagnosticSets. This prevents DiagnosticMap from type-checking reserved names. Ownership should also be split into GenericSets so that it is mergeable.

**Comment by eakuefner** [comment]
It seems like this is something else that would be worth smoke testing: if I add a new diagnostic type, merging two instances should not result in an UnmergeableDiagnosticSet.

**Comment by benshayden** [comment]
Yes! Let's do that right after splitting TelemetryInfo, Ownership, etc.

**Comment by benshayden** [comment]
How does this plan look?

1. [Reserve diagnostic names for fields of TelemetryInfo, DeviceInfo, BuildbotInfo, Ownership, RevisionInfo](https://codereview.chromium.org/2979723002)
1. Move reserved infos to a python module to facilitate splitting diagnostics to separate files #3613 
1. Maybe refactor reserved infos to instances of a class with methods
1. Add unindexed name to SparseDiagnostic
1. @phrmsilva Add device info GenericSets, update dashboard
1. @phrmsilva Add HistogramGroupings for device info GenericSets
1. @phrmsilva Delete DeviceInfo and its reserved name from both js and python, update documentation
1. @phrmsilva Split Ownership fields to GenericSets, update dashboard, change owners reserved info type from Ownership to GenericSet
1. @phrmsilva Delete Ownership from both js and python, update documentation
1. Build HistogramGroupings from tags (including `key:value` tags if storyGroupingKeys are still a thing)
1. Transform storyGroupingKeys and TIR to tags in metricMapFunction
1. Change dashboard to use reserved names/infos instead of reserved types including new names
1. Split TelemetryInfo fields to GenericSets/DateRanges, update chart json converter, compare_samples, and HistogramGroupings
1. Delete TelemetryInfo and associated reserved names from both js and python, update documentation

**Comment by eakuefner** [comment]
:+1:

**Comment by simonhatch** [comment]
Quick question, are we going to end up with a lot more suite/histogram sparse diagnostics this way or will a lot of these become dense?

**Comment by eakuefner** [comment]
I count 12 from Ben's CL https://codereview.chromium.org/2979723002:
  * Architectures
  * Benchmarks
  * Bots
  * Bug components
  * GPUs (GPU infos)
  * Masters
  * Memory amounts (system)
  * OS names
  * OS versions
  * Stories
  * Story tags
  * Owner emails

I don't think this is many more than we already had. Many of these will be stored at the suite level, with only stories and story tags, which would have been fields of TelemetryInfo, being stored at the histogram level. So my intuition is that it doesn't throw the math much, if that's the concern here.
